### PR TITLE
DPL-954: colour code wells on arraying page

### LINKED ISSUE #1427: DPL-954 Colour code wells on arraying page
**User story**
As a user of the scRNA Core Cell Extraction pipeline, when I am scanning LRC Blood Aliquot tubes into positions on the LRC Blood Bank plate (sample arraying page), I would like the wells to be colour coded according to the sample that they contain, to help avoid mistakes.

**Who are the primary contacts for this story**
Abby, Katy

**Who is the nominated tester for UAT**
Abby

**Acceptance criteria**
On the sample arraying screen (transfer from LRC Blood Aliquot to LRC Blood Bank, in the scRNA Core Cell Extraction pipeline):
- [ ] Wells are coloured the same as other wells that contain the same Sequencescape sample record, and differently from wells that contain a different sample.
- [ ] Colours that mean other things already are avoided (e.g. red is used in the well-failing feature)
- [ ] If this functionality bleeds over to the similar Cardinal screens I don't think this will matter (probably useful for them too) - but probably not worth trying to do this on purpose as that project is not very active any more. Make sure it doesn't break them though.

**References**
This is an enhancement to the original story https://github.com/sanger/limber/issues/1310

### PR DESCRIPTION
**Pull request by StephenHulme**
Closes #1427 

#### Changes proposed in this pull request

- Small UI touchups
- Add header showing child and parent purposes
- Repair a malform DOM element id
- Colour tubes based on machine barcode
- Show well tooltip with well position and human barcode
- Update tests
- Builds onto #1463 

Screenshots:

Screenshot 2024-02-13 at 14 27 31

Screenshot 2024-02-20 at 14 24 27

#### Deployment Notes

This should be tested against Integration Suite in UAT to confirm no breaking changes are introduced.

#### Instructions for Reviewers

_[All PRs] - Confirm PR template filled_  
_[Feature Branches] - Review code_  
_[Production Merges to `main`]_  
 &nbsp; &nbsp; \- _Check story numbers included_  
 &nbsp; &nbsp; \- _Check for debug code_  
 &nbsp; &nbsp; \- _Check version_

### CODE REVIEW

File: app/javascript/multi-stamp-tubes/components/MultiStampTubes.spec.js [hunk 1]
```diff
@@ -268,8 +268,8 @@ describe('MultiStampTubes', () => {
     expect(wrapper.vm.targetWells).toEqual({
       A1: { pool_index: 1 },
       B1: { pool_index: 2 },
-      C1: { pool_index: 3 },
-      D1: { pool_index: 4 },
+      C1: { pool_index: 2 },
+      D1: { pool_index: 3 },
     })
   })
```
Comment (andrewsparkes) [inline app/javascript/multi-stamp-tubes/components/MultiStampTubes.spec.js:274]: Wondered if there were tests for the individual Vue components.

Also wondered what happens when they remove the last instance of a barcode from the scan boxes, does it free up that colour index to be used for the next new unique barcode?

File: app/javascript/multi-stamp-tubes/components/MultiStampTubes.spec.js [hunk 2]
```diff
@@ -268,8 +268,8 @@ describe('MultiStampTubes', () => {
     expect(wrapper.vm.targetWells).toEqual({
       A1: { pool_index: 1 },
```
Comment (andrewsparkes) [inline app/javascript/multi-stamp-tubes/components/MultiStampTubes.spec.js:269]: pool_index is odd (existing) naming, we're not pooling here, it's just an index for the unique barcode

File: app/javascript/multi-stamp-tubes/components/MultiStampTubes.spec.js (hunk 1 above)
Comment (andrewsparkes) [reply app/javascript/multi-stamp-tubes/components/MultiStampTubes.spec.js:274]: It probably shouldn't right, if that one they delete was in the middle of the colour list you don't want complications with latter ones being re-coloured.

**Review by andrewsparkes — CHANGES_REQUESTED** [review]
Code looks good (although I'm far from a JS expert)
I asked if there were Vue component unit tests, and if you could maybe test the scenarios where they delete/change barcodes.

**Review by BenTopping — COMMENTED** [review]
Haven't tested locally but changes seem sensible.

File: app/javascript/multi-stamp-tubes/components/MultiStampTubes.vue [hunk 3]
```diff
@@ -204,13 +212,14 @@ export default {
...
+      return this.validTransfers.reduce((acc, transfer) => {
+        const tubeIndex = transfer.tubeObj.index
+        acc[transfer.targetWell] = {
+          colour_index: this.colourIndex(tubeIndex),
+          human_barcode: this.tubes[tubeIndex].labware.labware_barcode.human_barcode,
         }
-      }
-      return wells
+        return acc
+      }, {})
```
Comment (KatyTaylor) [inline app/javascript/multi-stamp-tubes/components/MultiStampTubes.vue:222]: Looks like you've changed the structure of the return value here, so could you update the comment above the function to reflect that?

File: app/javascript/multi-stamp-tubes/components/MultiStampTubes.vue [hunk 4]
```diff
@@ -204,13 +212,14 @@ export default {
...
     targetWells() {
-      const wells = {}
-      for (let i = 0; i < this.validTransfers.length; i++) {
-        wells[this.validTransfers[i].targetWell] = {
-          pool_index: this.validTransfers[i].tubeObj.index + 1,
+      return this.validTransfers.reduce((acc, transfer) => {
+        const tubeIndex = transfer.tubeObj.index
+        acc[transfer.targetWell] = {
+          colour_index: this.colourIndex(tubeIndex),
+          human_barcode: this.tubes[tubeIndex].labware.labware_barcode.human_barcode,
```
Comment (KatyTaylor) [inline app/javascript/multi-stamp-tubes/components/MultiStampTubes.vue:219]:
```suggestion
          human_barcode: transfer.tubeObj.tube.labware_barcode.human_barcode
```

Looks to me like the above should work, and is neater to get it from the variable `transfer` I think than having to use `this.tubes`?

File: app/javascript/multi-stamp-tubes/components/MultiStampTubes.vue
Comment (KatyTaylor) [inline app/javascript/multi-stamp-tubes/components/MultiStampTubes.vue:1]: I've had some trouble working out the data structure we're expecting for inputs and outputs of certain functions. Could we add docs, for example the following above the `validTransfers` function?

```
    //  validTransfers returns an array with the following structure:
    //
    //  [
    //    { tubeObj: { index: 0, tube: {...} }, targetWell: 'A1' },
    //    { tubeObj: { index: 1, tube: {...} }, targetWell: 'A2' },
    //    ...etc...
    //  ]
    //
```

File: app/javascript/multi-stamp-tubes/components/MultiStampTubes.vue [hunk 5]
```diff
@@ -236,6 +245,23 @@ export default {
     wellIndexToName(index) {
       return indexToName(index, this.targetRowsNumber)
     },
+    colourIndex(tubeIndex) {
```
Comment (KatyTaylor) [inline app/javascript/multi-stamp-tubes/components/MultiStampTubes.vue:248]: Can we have docs, something like:
```
    // Determines the colour of the tube based on its barcode, 
    // where tubes with the same barcode have the same colour.
    // Returns an integer that is used elsewhere to build the colour class name (see colours.css).
    // Retruns -1 if the tube is not valid.
```

File: app/javascript/multi-stamp-tubes/components/MultiStampTubes.vue (hunk 3 above)
Comment (KatyTaylor) [reply app/javascript/multi-stamp-tubes/components/MultiStampTubes.vue:222]: Thanks for renaming to colour_index, that seems to make a lot more sense.

**Review by KatyTaylor — CHANGES_REQUESTED** [review]
Seems to work well locally for me apart from the one bug I mentioned to you. I've left a few comments - mostly suggestions for adding docs. I find the existing component code quite hard to follow (not your fault). Thanks for doing all the tests.

File: app/javascript/multi-stamp-tubes/components/MultiStampTubes.spec.js (hunk 1 above)
Comment (andrewsparkes) [reply app/javascript/multi-stamp-tubes/components/MultiStampTubes.spec.js:368]: _Deleting the barcode will cause all other codes to be shifted up by one._

As in it shifts them up into new rack locations? I think that would be bad if that's what you mean, as they are probably physically placing the tubes in a rack as they scan them.

Comment (andrewsparkes) [reply app/javascript/multi-stamp-tubes/components/MultiStampTubes.spec.js:368]: If you mean just shifted in colour index then that's acceptable I think. The colours are just a visual aid to see where the replicates are.

**Review by andrewsparkes — APPROVED** [review]
Looks good. Helpful comments and better tests now! Thx.

File: app/javascript/multi-stamp-tubes/components/MultiStampTubes.spec.js (hunk 1 above)
Comment (KatyTaylor) [reply app/javascript/multi-stamp-tubes/components/MultiStampTubes.spec.js:368]: I noticed that functionality too, the colours changing when you remove a barcode. I thought it was ok.

**Review by KatyTaylor — APPROVED** [review]
Great, thanks for adding those extra comments.

**Review by BenTopping — APPROVED** [review]
